File: problems/0010-regular-expression-matching/solution.cpp
#include <vector>
#include <string>

class Solution {
public:
    bool isMatch(const std::string& s, const std::string& p) {
        /*
        Problem 0010 – Regular Expression Matching.

        dp[i][j] == true ⇔ the suffix s[i:] matches the suffix p[j:].
        We fill the table from the end of the strings toward the front.

        Time Complexity  : O(m × n),  m = |s|, n = |p|.
        Space Complexity : O(m × n).
        */

        int m = static_cast<int>(s.size());
        int n = static_cast<int>(p.size());

        std::vector<std::vector<bool>> dp(m + 1, std::vector<bool>(n + 1, false));
        dp[m][n] = true;                                     // empty string vs. empty pattern.

        for (int i = m; i >= 0; --i) {
            for (int j = n - 1; j >= 0; --j) {               // j == n already initialised.
                bool firstMatch = (i < m) &&
                                   (p[j] == s[i] || p[j] == '.');

                if (j + 1 < n && p[j + 1] == '*') {          // pattern has “x*”.
                    bool zeroOcc  = dp[i][j + 2];             // skip “x*”.
                    bool oneOrMore = firstMatch && dp[i + 1][j]; // consume one char of s.
                    dp[i][j] = zeroOcc || oneOrMore;
                } else {                                      // ordinary single-character match.
                    dp[i][j] = firstMatch && dp[i + 1][j + 1];
                }
            }
        }
        return dp[0][0];
    }
};

File: problems/0010-regular-expression-matching/solution.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        """
        Problem: 0010 - Regular Expression Matching.

        Determines whether the input string `s` matches the pattern `p`,
        where:
            - `.` matches any single character.
            - `*` matches zero or more of the preceding character.

        The matching must cover the entire string `s`, not just a substring.

        Approach:
        We use bottom-up dynamic programming. Let dp[i][j] be True if s[i:] matches p[j:].

        Time Complexity: O(m * n), where m = len(s), n = len(p).
        Space Complexity: O(m * n).
        """

        m = len(s)
        n = len(p)

        # dp[i][j] means: does s[i:] match p[j:]?
        dp = [[False] * (n + 1) for _ in range(m + 1)]

        # Base case: both s and p are empty.
        dp[m][n] = True

        # Fill the table in reverse order.
        for i in range(m, -1, -1):
            for j in range(n - 1, -1, -1):

                # Check if the first characters match (if in bounds),
                # or if pattern has a '.' wildcard.
                first_match = i < m and (p[j] == s[i] or p[j] == '.')

                # Handle the '*' operator, which applies to p[j + 1].
                if j + 1 < n and p[j + 1] == '*':
                    # Case 1: '*' represents zero occurrences of p[j].
                    # We skip over the 'x*' (move two steps in pattern).
                    zero_occurrence = dp[i][j + 2]

                    # Case 2: '*' represents one or more of p[j].
                    # Only possible if s[i] matches p[j]; then consume s[i] and stay at p[j].
                    one_or_more = first_match and dp[i + 1][j]

                    dp[i][j] = zero_occurrence or one_or_more

                else:
                    # No '*': just require current characters to match and recurse.
                    dp[i][j] = first_match and dp[i + 1][j + 1]

        # Final result: does s[0:] match p[0:]?
        return dp[0][0]

